dropshot/examples/self-referential.rs

```diff
@@ -0,0 +1,101 @@
+// Copyright 2020 Oxide Computer Company
+
+use dropshot::endpoint;
+use dropshot::ApiDescription;
+use dropshot::ConfigDropshot;
+use dropshot::ConfigLogging;
+use dropshot::ConfigLoggingLevel;
+use dropshot::HttpError;
+use dropshot::HttpResponseOk;
+use dropshot::HttpServerStarter;
+use dropshot::RequestContext;
+use schemars::JsonSchema;
+use serde::Deserialize;
+use serde::Serialize;
+use std::sync::atomic::AtomicU64;
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+
+#[tokio::main]
+async fn main() -> Result<(), String> {
+    let config_dropshot: ConfigDropshot = Default::default();
+    let config_logging =
+        ConfigLogging::StderrTerminal { level: ConfigLoggingLevel::Info };
+    let log = config_logging
+        .to_logger("example-self-referential")
+        .map_err(|error| format!("failed to create logger: {}", error))?;
+
+    let mut api = ApiDescription::new();
+    api.register(example_api_get_counter).unwrap();
+
+    let api_context = Arc::new(ExampleContext::new());
+
+    // Set up the server.
+    //
+    // Note that we split the "closer" from the "server", so we can continue
+    // to reference the server while simultaneously waiting for it to shut down.
+    let server =
+        HttpServerStarter::new(&config_dropshot, api, api_context, &log)
+            .map_err(|error| format!("failed to create server: {}", error))?
+            .start();
+    let (server, closer) = server.get_closer();
+    let (server, joiner) = server.get_joiner();
+
+    tokio::task::spawn(async move {
+        loop {
+            // We can access the server's data while simultaneously awaiting
+            // its termination.
+            let value =
+                server.app_private().counter.fetch_add(1, Ordering::SeqCst);
+            const TIMEOUT: u64 = 5;
+            if value >= TIMEOUT {
+                println!("Terminating now");
+                break;
+            } else {
+                println!("Terminating in {} seconds", TIMEOUT - value);
+                tokio::time::sleep(tokio::time::Duration::from_secs(1)).await;
+            }
+        }
+        // Once the timeout has been reached, we stop the server ourselves.
+        closer.close().await.unwrap();
+    });
+
+    // Wait for the server to stop.
+    joiner.await
```

Comment:
Not that we're joining on the `joiner` here. It's not possible to call `server.await` after calling `server.get_joiner`, but doing so let's the "server" and "joiner" have different lifetimes.